File: src/main/modelCatalog.ts
import type {
  AgentRole,
  AvailableModel,
  DetectedModelOption,
  DetectedProviderProfile,
  ProviderKind
} from './types'

type AvailabilityStatus = AvailableModel['availabilityStatus']
type BillingKind = AvailableModel['billingKind']

const PROVIDER_LABELS: Record<ProviderKind, string> = {
  opencode: 'OpenCode',
  codex: 'Codex',
  claude: 'Claude Code',
  gemini: 'Gemini CLI'
}

const SOURCE_PROVIDER_LABELS: Record<string, string> = {
  openai: 'OpenAI',
  anthropic: 'Anthropic',
  google: 'Google',
  openrouter: 'OpenRouter',
  groq: 'Groq',
  mistral: 'Mistral'
}

function capitalizeWord(word: string): string {
  if (!word) return word
  return word[0].toUpperCase() + word.slice(1)
}

function formatWords(value: string): string {
  return value
    .split(/[-_/\s]+/)
    .filter(Boolean)
    .map((word) => capitalizeWord(word))
    .join(' ')
}

function getSourceProviderLabel(sourceProvider: string | undefined, kind: ProviderKind): string {
  if (sourceProvider) {
    return SOURCE_PROVIDER_LABELS[sourceProvider] ?? formatWords(sourceProvider)
  }

  if (kind === 'codex') return 'OpenAI'
  if (kind === 'claude') return 'Anthropic'
  if (kind === 'gemini') return 'Google'
  return 'Configured Provider'
}

function getRecommendedRoles(modelId: string): AgentRole[] {
  const normalized = modelId.toLowerCase()

  if (normalized.includes('mini') || normalized.includes('flash') || normalized.includes('haiku')) {
    return ['executor']
  }

  if (
    normalized.includes('pro') ||
    normalized.includes('sonnet') ||
    normalized.includes('opus') ||
    normalized.includes('o1')
  ) {
    return ['mentor']
  }

  return ['mentor', 'executor']
}

function getAvailability(
  profile: DetectedProviderProfile,
  model: DetectedModelOption
): { status: AvailabilityStatus; reason?: string; available: boolean } {
  const providerLabel = PROVIDER_LABELS[profile.kind]

  if (!profile.installed) {
    return {
      status: 'cli-missing',
      reason: `${providerLabel} CLI is not installed`,
      available: false
    }
  }

  if (!profile.authenticated) {
    return {
      status: 'auth-missing',
      reason: `${providerLabel} is not signed in`,
      available: false
    }
  }

  if (!profile.runnable || !model.runnable || !model.supportsPairExecution) {
    return {
      status: 'runtime-unsupported',
      reason: `${providerLabel} is detected, but pair execution is not yet supported`,
      available: false
    }
  }

  return {
    status: 'ready',
    available: true
  }
}

function getPlanLabel(profile: DetectedProviderProfile): string | undefined {
  const normalized = String(profile.subscriptionLabel || '')
    .trim()
    .toLowerCase()

  if (!normalized || normalized === 'authenticated' || normalized === 'provider-backed') {
    return undefined
  }

  if (
    profile.kind === 'codex' &&
    ['free', 'plus', 'pro', 'team', 'enterprise'].includes(normalized)
  ) {
    return `ChatGPT ${formatWords(normalized)}`
  }

  if (normalized === 'subscription-backed') {
    if (profile.kind === 'claude') return 'Claude subscription'
    if (profile.kind === 'gemini') return 'Google AI plan'
    if (profile.kind === 'codex') return 'ChatGPT account'
  }

  return formatWords(normalized)
}

function getBilling(
  profile: DetectedProviderProfile,
  sourceProviderLabel: string
): { kind: BillingKind; label: string; accessLabel: string; planLabel?: string } {
  const planLabel = getPlanLabel(profile)

  if (profile.kind === 'opencode') {
    return {
      kind: 'byok',
      label: 'Pay as you go',
      accessLabel: `${sourceProviderLabel} API key`,
      planLabel: 'BYOK'
    }
  }

  if (profile.kind === 'codex') {
    if (planLabel) {
      return {
        kind: 'plan',
        label: 'Included with plan',
        accessLabel: 'ChatGPT plan',
        planLabel
      }
    }

    if (profile.authenticated) {
      return {
        kind: 'plan',
        label: 'Included with account',
        accessLabel: 'Codex login'
      }
    }
  }

  if (profile.kind === 'claude') {
    return {
      kind: 'plan',
      label: planLabel ? 'Included with plan' : 'Requires account access',
      accessLabel: 'Claude Code login',
      planLabel
    }
  }

  if (profile.kind === 'gemini') {
    return {
      kind: 'plan',
      label: planLabel ? 'Included with plan' : 'Requires Google account',
      accessLabel: 'Google account',
      planLabel
    }
  }

  return {
    kind: 'unknown',
    label: 'Unknown billing',
    accessLabel: sourceProviderLabel
  }
}

export function buildModelCatalog(profiles: DetectedProviderProfile[]): AvailableModel[] {
  return profiles
    .flatMap((profile) => {
      const providerLabel = PROVIDER_LABELS[profile.kind]

      return profile.currentModels.map((model) => {
        const sourceProviderLabel = getSourceProviderLabel(model.sourceProvider, profile.kind)
        const availability = getAvailability(profile, model)
        const billing = getBilling(profile, sourceProviderLabel)

        return {
          provider: profile.kind,
          modelId: model.modelId,
          displayName: model.displayName,
          available: availability.available,
          providerLabel,
          sourceProvider: model.sourceProvider,
          sourceProviderLabel,
          billingKind: billing.kind,
          billingLabel: billing.label,
          accessLabel: billing.accessLabel,
          planLabel: billing.planLabel,
          availabilityStatus: availability.status,
          availabilityReason: availability.reason,
          supportsPairExecution: model.supportsPairExecution,
          recommendedRoles: getRecommendedRoles(model.modelId)
        } satisfies AvailableModel
      })
    })
    .sort((left, right) => {
      if (left.available !== right.available) {
        return left.available ? -1 : 1
      }

      if (left.providerLabel !== right.providerLabel) {
        return left.providerLabel.localeCompare(right.providerLabel)
      }

      return left.displayName.localeCompare(right.displayName)
    })
}

File: src-tauri/src/model_catalog.rs
use crate::provider_registry::{DetectedProviderProfile, ProviderKind};
use serde::Serialize;

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct AvailableModel {
    pub provider: ProviderKind,
    pub model_id: String,
    pub display_name: String,
    pub available: bool,
    pub provider_label: String,
    pub source_provider: Option<String>,
    pub source_provider_label: String,
    pub billing_kind: String,
    pub billing_label: String,
    pub access_label: String,
    pub plan_label: Option<String>,
    pub availability_status: String,
    pub availability_reason: Option<String>,
    pub supports_pair_execution: bool,
    pub recommended_roles: Vec<String>,
}

pub struct ModelCatalog;

impl ModelCatalog {
    pub fn build_catalog(profiles: Vec<DetectedProviderProfile>) -> Vec<AvailableModel> {
        let mut catalog = Vec::new();

        for profile in profiles {
            let provider_label = match profile.kind {
                ProviderKind::Opencode => "OpenCode",
                ProviderKind::Codex => "Codex",
                ProviderKind::Claude => "Claude Code",
                ProviderKind::Gemini => "Gemini CLI",
            };

            for model in profile.current_models {
                let source_provider_label = match model.source_provider.as_deref() {
                    Some("openai") => "OpenAI",
                    Some("anthropic") => "Anthropic",
                    Some("google") => "Google",
                    _ => "Configured Provider",
                };

                let (status, reason, available) = if !profile.installed {
                    ("cli-missing".to_string(), Some(format!("{} CLI is not installed", provider_label)), false)
                } else if !profile.authenticated {
                    ("auth-missing".to_string(), Some(format!("{} is not signed in", provider_label)), false)
                } else if !profile.runnable || !model.runnable || !model.supports_pair_execution {
                    ("runtime-unsupported".to_string(), Some(format!("{} is detected, but pair execution is not yet supported", provider_label)), false)
                } else {
                    ("ready".to_string(), None, true)
                };

                let billing_kind = match profile.kind {
                    ProviderKind::Opencode => "byok",
                    _ => "plan",
                };

                let billing_label = match profile.kind {
                    ProviderKind::Opencode => "Pay as you go",
                    _ => "Included with plan",
                };

                let access_label = match profile.kind {
                    ProviderKind::Opencode => format!("{} API key", source_provider_label),
                    ProviderKind::Codex => "ChatGPT plan".into(),
                    ProviderKind::Claude => "Claude Code login".into(),
                    ProviderKind::Gemini => "Google account".into(),
                };

                catalog.push(AvailableModel {
                    provider: profile.kind,
                    model_id: model.model_id.clone(),
                    display_name: model.display_name,
                    available,
                    provider_label: provider_label.to_string(),
                    source_provider: model.source_provider,
                    source_provider_label: source_provider_label.to_string(),
                    billing_kind: billing_kind.to_string(),
                    billing_label: billing_label.to_string(),
                    access_label,
                    plan_label: Some(profile.subscription_label.clone()),
                    availability_status: status,
                    availability_reason: reason,
                    supports_pair_execution: model.supports_pair_execution,
                    recommended_roles: vec!["mentor".into(), "executor".into()], // Simplified
                });
            }
        }

        catalog.sort_by(|a, b| {
            if a.available != b.available {
                return b.available.cmp(&a.available);
            }
            a.provider_label.cmp(&b.provider_label)
        });

        catalog
    }
}
